peps/pep-0734.rst

```diff
@@ -0,0 +1,800 @@
+it will be destroyed as soon as all ``Interpreter`` objects have been
+deleted.
+
+Attributes and methods:
+
+* ``id``
+      (read-only) The target interpreter's ``InterpreterID`` object.
+      It is an int-like object with some internal bookkeeping.
+
+* ``__hash__()``
+      Returns the hash of the interpreter's ``id``.  This is the same
+      as the hash of the ID's integer value.
+
+* ``__eq__(other)``
+      Returns ``other is self``.
+
+* ``is_running() -> bool``
+      Returns ``True`` if the interpreter is currently executing code
+      in its ``__main__`` module.  This excludes sub-threads.
+
+      It refers only to if there is a thread
+      running a script (code) in the interpreter's ``__main__`` module.
+      That basically means whether or not ``Interpreter.exec()`` is
+      running in some thread.  Code running in sub-threads is ignored.
+
+* ``prepare___main__(**kwargs)``
+      Bind one or more objects in the interpreter's ``__main__`` module.
+
+      The keyword argument names will be used as the attribute names.
+      The values will be bound as new objects, though exactly equivalent
+      to the original.  Only objects specifically supported for passing
+      between interpreters are allowed.  See `Shareable Objects`_.
+
+      ``prepare___main__()`` is helpful for initializing the
+      globals for an interpreter before running code in it.
+
+* ``exec(code, /)``
+      Run the given source code in the interpreter
+      (in the current thread).
+
+      This is essentially equivalent to switching to this interpreter
+      in the current thread and then calling the builtin ``exec()``
+      using this interpreter's ``__main__`` module's ``__dict__`` as
+      the globals and locals.
+
+      ``exec()`` does not reset the interpreter's state nor
+      the ``__main__`` module, neither before nor after, so each
+      successive call picks up where the last one left off.  This can
+      be useful for running some code to initialize an interpreter
+      (e.g. with imports) before later performing some repeated task.
+
+Comparison with builtins.exec()
+-------------------------------
+
+``Interpreter.exec()`` is essentially the same as the builtin
+``exec()``, except it targets a different interpreter, using that
+interpreter's distinct runtime state.
+
+Here are the relevant characteristics of the builtin ``exec()``,
+for comparison:
+
+* It runs in the current OS thread and pauses whatever was running there,
+  which resumes when ``exec()`` finishes.  No other threads are affected.
+  (To avoid pausing the current thread, run ``exec()``
+  in a ``threading.Thread``.)
+* It executes against a namespace, by default the ``__dict__`` of the
+  current module (i.e. ``globals()``).  It uses that namespace as-is
+  and does not clear it before or after.
+* When code is run from the command-line (e.g. ``-m``) or the REPL,
+  the "current" module is always the ``__main__`` module
+  of the target (main) interpreter.
```

Comment:
How is this relevant to Interpreter.exec()?

Reply:
That feels like a stretch -- I wouldn't say that `exec()` is like `python -c` either. (E.g., `-c` prints the traceback, but `exec()` doesn't -- the traceback printing is in a different part of the functionality.)

Rather than trying to use analogies, I think it would work better if you just specified the intended behavior carefully.

Reply:
> That feels like a stretch -- I wouldn't say that `exec()` is like `python -c` either. (E.g., `-c` prints the traceback, but `exec()` doesn't -- the traceback printing is in a different part of the functionality.)

FWIW, I am moving this up to the Introduction and converting this specific bullet into a distinct, brief discussion of how the different CLI options relate to the relevant C API functions (e.g. `PyRun_String()`, `PyEval_EvalCode()`).  That's in the interest of identifying how `exec()`, `Interpreter.exec()`, and the CLI are related at a fundamental level.

> Rather than trying to use analogies, I think it would work better if you just specified the intended behavior carefully.

Fair enough.  I'll clean that up.